File: W6D1/array_exercises.js
Array.prototype.my_includes = function(val) {
  for(var i = 0; i < this.length; i++) {
    if(this[i] === val) {
      return true;
    }
  };
  return false;
}

Array.prototype.my_uniq = function() {
  var uniq_array = [];
  for(var i = 0; i < this.length; i++) {
    if(!uniq_array.my_includes(this[i])) {
      uniq_array.push(this[i]);
    }
  };
  return uniq_array;
}

console.log([1, 2, 1, 3, 3].my_uniq());

Array.prototype.two_sum = function() {
  var two_sum_array = []
  for(var i = 0; i < (this.length-1); i++) {
    for(var j = (i+1); j < this.length; j++) {
      if(this[i] + this[j] === 0) {
        two_sum_array.push([i, j]);
      }
    };
  };
  return two_sum_array;
}

console.log([-1, 0, 2, -2, 1].two_sum());

Array.prototype.my_transpose = function() {
  var transposed_array = []
  for(var i = 0; i < this.length; i++) {
    var transposed_row = []
    for(var j = 0; j < this[i].length; j++) {
      transposed_row.push(this[j][i]);
    };
    transposed_array.push(transposed_row);
  };
  return transposed_array;
}

var rows = [
    [0, 1, 2],
    [3, 4, 5],
    [6, 7, 8]
  ];

console.log(rows.my_transpose());

File: Week1/W1D1/array_exercises.rb
class Array
  def my_uniq
    result = []
    self.each do |val|
      result << val unless result.include?(val)
    end
    result
  end

  def two_sum
    result = []
    0.upto(self.length-1) do |x|
      (x+1).upto(self.length-1) do |y|
        result << [x, y] if self[x] + self[y] == 0
      end
    end
    result
  end
end

p [1, 2, 1, 3, 3].my_uniq
p [-1, 0, 2, -2, 1].two_sum


def my_transpose(arr)
  result = []
  0.upto(arr.length-1) do |col|
		new_row = []
		arr.each do |row|
			new_row << row[col]
		end
		result << new_row
	end
	result
end

rows = [
    [0, 1, 2],
    [3, 4, 5],
    [6, 7, 8]
	]

cols = [
    [0, 3, 6],
    [1, 4, 7],
    [2, 5, 8]
  ]

p my_transpose(rows)
p my_transpose(cols)

def stock_picker(stocks)
	max_ids = [0,0]
	stocks.each_with_index do |buy, buy_i|
		buy_i.upto(stocks.length-1) do |sell_i|
			sell = stocks[sell_i]
			if sell - buy > stocks[max_ids[1]] - stocks[max_ids[0]]
				max_ids = [buy_i, sell_i]
			end
		end
	end
	max_ids
end

p stock_picker([0, 20, 5, 8])
p stock_picker([5,10,100,20])